horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/RunServiceImpl.java

```diff
@@ -982,4 +979,131 @@ public QueryResult queryDataSet(Integer datasetId, String jsonpath, boolean arra
       }
       return result;
    }
+
+   private void transform(Run run, String role) {
+      if (run == null || role == null) {
+         log.errorf("Transformation parameters error: run %s, role %s", run, role);;
+      }
+      Collection<RunServiceImpl.RunSchema> runSchemas = getSchemas(run, role);
+      Test test = Test.findById(run.testid);
+      String data = run.data.toString();
+      JsonNode node = Util.toJsonNode(data);
+      List<JsonNode> nodesToProcess = new ArrayList<JsonNode>();
+
+      if (runSchemas.isEmpty()) {
+         new DataSet(Instant.now(), Instant.now(), "",
+            test.id, Util.toJsonNode(EMPTY_ARRAY), run, 0, run.owner, run.access)
+            .persist();
+      } else {
+         runSchemas.forEach( s -> {
+            if (hasMatch(s, node)) {
+               nodesToProcess.add(node);
+            }
+            Schema schema = Schema.find("uri = ?1", s.uri).singleResult();
+            List<Transformer> transformers = Transformer.find("schema_id = ?1", schema.id).list();
+                  new ArrayList<>(nodesToProcess.size());
+
+            int nodeSize = nodesToProcess.size();
+
+            if (nodeSize == 0) {
+               new DataSet(Instant.now(), Instant.now(), "",
+                     test.id, run.data, run, 0, run.owner, run.access)
+                     .persist();
+            } else {
+               if (transformers.isEmpty()) {
+                  new DataSet(Instant.now(), Instant.now(), "",
+                        test.id, run.data, run, 0, run.owner, run.access)
+                        .persist();
+               }
+               transformers.forEach(t -> {
+                  StringBuilder array = new StringBuilder("[");
+                  boolean atStart = true;
+                  boolean errors = false;
+                  for (int nodePos = 0; nodePos <= nodeSize; nodePos += 1 ) {
```

Comment:
Looks like you're applying the function to all values in the `data`; you need to go through `Transformer.extractors` to extract sub-nodes using the JSONPath query and *then* apply the function on the composition of results.

Also note that the function itself is optional (can be null) - if it is not defined the results will be used as `DataSet.data`.

Reply:
think I have those changes captured.

Reply:
ok